File: DSA/Cpp/src/dsa/lib/lists/linkedlists/ls.cpp
#include "ls.h"

namespace dsa::lib::lists::linkedlists {

LinkedList::LinkedList(ListNode *head) : Base(head) {}

// reverse the nodes before a certain node
ListNode *LinkedList::reverseBefore(ListNode *node) const { return reverseBefore1(this->head, node); }
ListNode *LinkedList::reverseBefore1(ListNode *head, ListNode *node) {
    ListNode *pre, *cur, *next;
    pre = node;
    cur = head;

    while (cur != node) {
        next = cur->next;
        cur->next = pre;
        pre = cur;
        cur = next;
    }
    return pre;
}
ListNode *LinkedList::reverse() const {
    ListNode *pre, *cur, *next;
    pre = nullptr;
    cur = this->head;

    while (cur) {
        next = cur->next;
        cur->next = pre;
        pre = cur;
        cur = next;
    }
    return pre;
}
// Reverse the nodes from *fnode* to *tnode*, does not include the *tnode* node
ListNode *LinkedList::reverseFromTo(ListNode *fnode, ListNode *tnode) const {
    return reverseFromTo1(this->head, fnode, tnode);
}
ListNode *LinkedList::reverseFromTo1(ListNode *head, ListNode *fnode, ListNode *tnode) {
    ListNode *cur = head;
    ListNode dummy;
    dummy.next = cur;
    ListNode *groupPre = &dummy;

    while (cur != fnode) {
        groupPre = cur;
        cur = cur->next;
    }

    groupPre->next = reverseBefore1(fnode, tnode);

    return dummy.next;
}

bool LinkedList::hasCycle() const {
    return hasCycle1(this->head);
}
/**
 * Rule:
 * Two Pointers
 * Fast and Slow Pointers
 * Tortoise and Hare Algorithm
 */
bool LinkedList::hasCycle1(ListNode *head) {
    ListNode *slow = head, *fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

ListNode *LinkedList::detectCycle() const {
    return detectCycle1(this->head);
}
// 2 pointers
// Fast and Slow pointers
ListNode *LinkedList::detectCycle1(ListNode *head) {
    ListNode *slow = head, *fast = head, *start = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) { // There is a cycle
            while (slow != start) {
                slow = slow->next;
                start = start->next;
            }
            return start;
        }
    }
    return nullptr; // There is no cycle
}
ListNode *LinkedList::detectCycle11(ListNode *head) {
    ListNode *slow = head, *fast = head;

    // Determine whether there's a cycle
    while (true) {
        if (not fast || not fast->next) return nullptr; // No cycle
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) break;
    }

    // Get the cycle entry node
    ListNode *start = head;
    while (start != slow) {
        start = start->next;
        slow = slow->next;
    }
    return start;
}

ListNode *LinkedList::swapPairs() {
    return swapPairs2(this->head);
}
//Iteration
ListNode *LinkedList::swapPairs1(ListNode *head) {
    ListNode dummy;
    dummy.next = head;

    ListNode *pre = &dummy;
    while (pre->next && pre->next->next) {
        ListNode *p = pre->next, *q = p->next;
//        swap
        pre->next = q;
        p->next = q->next;
        q->next = p;

        pre = pre->next->next;
    }

    return dummy.next;
}
// Recursion
ListNode *LinkedList::swapPairs2(ListNode *head) {
    if (head && head->next) {
        ListNode *newHead = head->next->next;
        ListNode *p = head, *q = p->next;
//        swap
        head = q;
        p->next = swapPairs2(newHead);
        q->next = p;
    }
    return head;
}
ListNode *LinkedList::swapPairs3(ListNode *head) {
    ListNode dummy;
    dummy.next = head;
    return swapPairsRecur(&dummy);
}
ListNode *LinkedList::swapPairsRecur(ListNode *prev) {
    if (prev->next && prev->next->next) {
        ListNode *p = prev->next, *q = p->next;
        prev->next = q;
        p->next = swapPairsRecur(q);
        q->next = p;
    }
    return prev->next;
}

ListNode *LinkedList::reverseKGroup(int k) {
    return reverseKGroup3(this->head, k);
}
// Iterative
// Rule: Cut the LinkedList into the small piece of individual LinkedLists
ListNode *LinkedList::reverseKGroup1(ListNode *head, int k) {
    ListNode dummy;
    dummy.next = head;

    ListNode *pre = &dummy, *cur = head;
    while (true) {
//        generate the nodes for a group
        for (int i = 0; i < k; i++) {
            if (!cur) return dummy.next;
            cur = cur->next;
        }

//        reverse the nodes in the group
//        connect it with group prev node(aka, last piece of LinkedList)
        pre->next = reverseBefore1(head, cur);
//        go to the next LinkedList
        pre = head;
        head = cur;
    }
}
// Rule: Treat the LinkedList as a whole, group the LinkedList nodes.
ListNode *LinkedList::reverseKGroup2(ListNode *head, int k) {
    ListNode *groupHead = head, *cur = head;
    while (true) {
        for (int i = 0; i < k; ++i) {
            if (!cur) return head;
            cur = cur->next;
        }
        head = reverseFromTo1(head, groupHead, cur);
        groupHead = cur;
    }
}
// Recursive
ListNode *LinkedList::reverseKGroup3(ListNode *head, int k) {
    ListNode *cur = head;
    for (int i = 0; i < k; ++i) {
        if (!cur) return head;
        cur = cur->next;
    }

    ListNode *newHead = reverseBefore1(head, cur);
    head->next = reverseKGroup3(cur, k);
    return newHead;
}

}  // namespace dsa::lib::lists::linkedlists

File: DSA/Python/src/dsa/lib/lists/linkedlists/ls.py
from typing import Set, Optional

from dsa.lib.lists.linkedlists.base import Base
from dsa.nodes import ListNode


class LinkedList(Base):
    def __init__(self, head: ListNode):
        super().__init__(head)

    # LeetCode206
    def reverse(self) -> ListNode:
        pre = None
        cur: ListNode = self.head
        next_: ListNode
        while cur:
            next_ = cur.next
            cur.next = pre
            pre, cur = cur, next_
        return pre

    def reverse_before(self, node: ListNode) -> ListNode:
        # Reverse the LinkedList nodes before a certain node, doesn't  reverse the certain node
        return self.reverse_before1(self.head, node)

    @staticmethod
    def reverse_before1(head: ListNode, node: ListNode) -> ListNode:
        pre: ListNode = node
        cur: ListNode = head
        next_: ListNode
        while cur is not node:
            next_ = cur.next
            cur.next = pre
            pre, cur = cur, next_
        return pre

    def reverse_from_to(self, fnode: ListNode, tnode: ListNode) -> ListNode:
        return self._reverse_from_to1(self.head, fnode, tnode)

    @classmethod
    def _reverse_from_to1(cls, head: ListNode, fnode: ListNode, tnode: ListNode) -> ListNode:
        # Reverse a group of nodes from fnode to tnode, doesn't reverse the tnode node.
        cur: ListNode = head
        dummy: ListNode = ListNode()
        dummy.next = cur
        group_pre: ListNode = dummy

        # 1. Get the group previous node
        while cur is not fnode:
            group_pre = cur
            cur = cur.next

        # 2. Reverse the linkedlist from the fnode to the tnode
        # 3. Concatenate the reversed group nodes to the main linkedlist
        group_pre.next = cls.reverse_before1(fnode, tnode)

        return dummy.next

    # LeetCode141
    def hasCycle(self) -> bool:
        return self._hasCycle1(self.head)

    # HashSet
    @staticmethod
    def _hasCycle1(head: ListNode) -> bool:
        found: Set[ListNode] = set()
        while head:
            if head in found: return True
            found.add(head)
            head = head.next
        return False

    # Fast and Slow pointers, or Tortoise and hare algorithm.
    @staticmethod
    def _hasCycle2(head: ListNode) -> bool:
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if fast is slow: return True
        return False

    # Pythonic
    @staticmethod
    def _hasCycle3(head: ListNode) -> bool:
        fast = slow = head
        while fast and fast.next:
            fast, slow = fast.next.next, slow.next
            if fast is slow: return True
        return False

    # LeetCode142
    def detectCycle(self) -> ListNode:
        return self._detectCycle2(self.head)

    # Hash Set. Hash Table
    @staticmethod
    def _detectCycle1(head: ListNode) -> Optional[ListNode]:
        found: Set[ListNode] = set()
        while head:
            if head in found: return head
            found.add(head)
            head = head.next
        return None

    # 2 pointers. Fast and Slow pointers, Tortoise and Hare algorithm
    # Pythonic
    @staticmethod
    def _detectCycle2(head: ListNode) -> Optional[ListNode]:
        slow = fast = start = head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow is fast:
                while slow is not start:
                    slow, start = slow.next, start.next
                return start
        return None

    @staticmethod
    def _detectCycle21(head: ListNode) -> Optional[ListNode]:
        slow: ListNode = head
        fast: ListNode = head

        # Determine whether there is a cycle
        while True:
            if not fast or not fast.next: return None
            slow = slow.next
            fast = fast.next.next
            if slow is fast: break

        # get the entry location node of the cycle
        # Here fast == slow
        start: ListNode = head
        while start is not slow:
            slow = slow.next
            start = start.next
        return start

    @staticmethod
    def _detectCycle22(head: ListNode) -> Optional[ListNode]:
        # 1. Determine whether there is a cycle
        slow: ListNode = head
        fast: ListNode = head
        start: ListNode = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:  # There is a cycle
                # 2. Get the cycle entry node
                while slow is not start:
                    slow = slow.next
                    start = start.next
                return start
        return None  # There is no cycle

    # LeetCode24
    def swapPairs(self) -> ListNode:
        return self._swapPairs2(self.head)

    # Iteration
    @staticmethod
    def _swapPairs1(head: ListNode) -> ListNode:
        dummy = ListNode()
        dummy.next = head

        prev: ListNode = dummy
        while prev.next and prev.next.next:
            p, q = prev.next, prev.next.next
            prev.next, p.next, q.next = q, q.next, p
            prev = prev.next.next

        return dummy.next

    # Recursion
    @classmethod
    def _swapPairs2(cls, head: ListNode) -> ListNode:
        if head and head.next:
            new_start = head.next.next
            p, q = head, head.next
            # swap
            head, p.next, q.next = q, cls._swapPairs2(new_start), p

        return head

    @staticmethod
    def _swapPairs3(head: ListNode) -> ListNode:
        def helper(prev: ListNode) -> ListNode:
            if prev.next and prev.next.next:
                p, q = prev.next, prev.next.next
                prev.next, p.next, q.next = q, helper(q), p
            return prev.next

        dummy = ListNode()
        dummy.next = head
        return helper(dummy)

    # LeetCode25
    def reverseKGroup(self, k: int) -> ListNode:
        return self._reverseKGroup3(self.head, k)

    # Iterative
    # Rule: Cut the LinkedList into the small piece of individual LinkedLists
    @classmethod
    def _reverseKGroup1(cls, head: ListNode, k: int) -> ListNode:
        # Initialize
        dummy: ListNode = ListNode()
        dummy.next = head
        pre, cur = dummy, head

        while True:
            for _ in range(k):
                if not cur: return dummy.next
                cur = cur.next

            pre.next = cls.reverse_before1(head, cur)
            # go to the next group
            pre, head = head, cur

    # Rule: Treat the LinkedList as a whole, group the LinkedList nodes.
    @classmethod
    def _reverseKGroup2(cls, head: ListNode, k: int) -> ListNode:
        group_head = cur = head
        while True:
            # generate the nodes for a group
            for _ in range(k):
                if not cur: return head
                cur = cur.next
            head = cls._reverse_from_to1(head, group_head, cur)
            group_head = cur

    # Recursion
    @classmethod
    def _reverseKGroup3(cls, head: ListNode, k: int) -> ListNode:
        cur: ListNode = head
        for _ in range(k):
            if not cur: return head
            cur = cur.next

        new_head: ListNode = cls.reverse_before1(head, cur)

        head.next = cls._reverseKGroup3(cur, k)
        return new_head
